src/data-workspace/category-combo-table/category-combo-table-header.js

```diff
@@ -0,0 +1,130 @@
+import PropTypes from 'prop-types'
+import React, { useMemo } from 'react'
+import { useMetadata, selectors } from '../../metadata/index.js'
+import { useActiveCell } from '../data-entry-cell/index.js'
+import styles from './category-combo-table.module.css'
+import { PaddingCell } from './padding-cell.js'
+import { TotalHeader } from './total-cells.js'
+
+// Computes the span and columns to render in each category-row
+// columns are category-options, and needs to be repeated in case of multiple categories
+const useCategoryColumns = (categories, numberOfCoCs) => {
+    const { data: metadata } = useMetadata()
+    return useMemo(() => {
+        let catColSpan = numberOfCoCs
+        return categories.map((c) => {
+            const categoryOptions = selectors.getCategoryOptionsByCategoryId(
+                metadata,
+                c.id
+            )
+            const nrOfOptions = c.categoryOptions.length
+            // catColSpan should always be equal to nrOfOptions in last iteration
+            // unless anomaly with categoryOptionCombo-generation server-side
+            if (nrOfOptions > 0 && catColSpan >= nrOfOptions) {
+                // calculate colSpan for current options
+                // this is the span for each option, not the "total" span of the row
+                catColSpan = catColSpan / nrOfOptions
+                // when table have multiple categories, options need to be repeated for each disaggregation "above" current-category
+                const repeat = numberOfCoCs / (catColSpan * nrOfOptions)
+
+                const columnsToRender = new Array(repeat)
+                    .fill(0)
+                    .flatMap(() => categoryOptions)
+
+                return {
+                    span: catColSpan,
+                    columns: columnsToRender,
+                    category: c,
+                }
+            } else {
+                console.warn(
+                    `Category ${c.displayFormName} malformed. Number of options: ${nrOfOptions}, span: ${catColSpan}`
+                )
+            }
+            return c
+        })
+    }, [metadata, categories, numberOfCoCs])
+}
+
+export const CategoryComboTableHeader = ({
+    renderRowTotals,
+    paddingCells,
+    categoryOptionCombos,
+    categories,
+    checkTableActive,
+}) => {
+    const { deId: activeDeId, cocId: activeCocId } = useActiveCell()
+
+    const rowToColumnsMap = useCategoryColumns(
+        categories,
+        categoryOptionCombos.length
+    )
+
+    // Find if this column header includes the active cell's column
+    const isHeaderActive = (headerIdx, headerColSpan) => {
+        const activeCellColIdx = categoryOptionCombos.findIndex(
+            (coc) => activeCocId === coc.id
+        )
+        const idxDiff = activeCellColIdx - headerIdx * headerColSpan
+        return (
+            checkTableActive(activeDeId) &&
+            idxDiff < headerColSpan &&
+            idxDiff >= 0
+        )
+    }
```

Comment:
> I have one question for my own understanding: Why is using two providers (one "getter provider" and one "setter provider") functionally different from using a single provider? As I understanding it, it does seem to make a different, I just don't understand why yet. Could you explain that?


It's because when you are using context, you subscribe to the entire value of the context, and when it changes all subscribers re-render. So when the "item" part of the context-changes, all `EntryFieldInput`-components will re-render, even though they just need the "setItem"-part of the context. There's no way to choose to just subscribe to the setter-part. That's why a separate provider with just the setter-function makes it possible for this component to not re-render, they don't care about the actual value.

> I'm generally not a friend of abbreviations. It makes code more difficult to understand, in some cases it can even be ambiguous. Here it saves 2 characters (idx -> index) and at first I thought this is a weird plural of id). It also doesn't add any value other than saving typing time.


This is code that has been there before, I just moved it, and generally I'm not  a fan of refactoring others people's code, especially variable names when they're already in the codebase. And `idx` is a pretty usual abbreviation for index in programming, no? Is `diff` also too much abbreviation? Should it be called `indexDifference`? I think abbreviations have their place, and sometimes expanding everything can make it harder to keep track of variables etc. 


> Same applied to the map over rowToColumnsMap below, where sometimes it's colInfoIndex (column abbreviated, but Index is not) and columnIndex (where neither is abbreviated)


I mean, naming can be argued back and forth for hours. I agree these are not consistent but at some point the names would get too long. I think most understand what `colInfoIndex` means in this context.

Reply:
## Question about the two providers

Thanks! It makes complete sense to me now :)

## Abbreviations

> This is code that has been there before, I just moved it

Did mean to say that this is your fault or something like that! Just commented on the code I saw :)

---

> I'm not a fan of refactoring others people's code, especially variable names when they're already in the codebase.

I wouldn't count un-abbreviating as refactoring, I don't think we have to be that sensitive when refactoring. Of course if you're worried that this is important to original author, you can request a review from that person and I'd even encourage you to do that!

---

> And idx is a pretty usual abbreviation for index in programming, no?

Despite "refactoring", what's the advantage of using the abbreviation other than saving 2 characters? There's a disadvantage which is worse in relation to saving two character: _Understandability_. Yes `idx` is probably used in many places, but the fact that it confused me at first already means there are "developer bubbles" which haven't seen it, so I'm still wondering: What's the advantage here?

---

> I think abbreviations have their place, and sometimes expanding everything can make it harder to keep track of variables etc.

Yes, you're right, I don't disagree. `id` is an abbreviation in itself, but it's so common that it's sort of its own word already. I'd argue you can say the same about `diff`. To me these are different from `idx` as that one is less common (I actually have seen `i` more often, which I'm actually OK with as that's as common as `id`), it also looks very similar to `ids`, which makes it mentally more taxing.

> expanding everything can make it harder to keep track of variables etc.

Coming back to this and `idx` specifically. I don't see why `index` makes it harder to keep track of variables? As mentioned, I wouldn't un-abbreviate everything (e.g. `de`, as it has an established meaning in the context of dhis2)

Reply:
I see that you marked this as old, but will just follow up with some minor comments. 

> Did mean to say that this is your fault or something like that! Just commented on the code I saw :)

I totally understand! Just wanted to make it clear and was mostly and argument for why I didn't want to touch too much of that code.


> I wouldn't count un-abbreviating as refactoring, I don't think we have to be that sensitive when refactoring. Of course if you're worried that this is important to original author, you can request a review from that person and I'd even encourage you to do that!

Hm... I think it's definitely refactoring. I might be too strict about my thoughts on refactoring and stuff, and probably need to learn to be a bit more lenient. Not sure what it is, but it's probably a bit about the respect of the other's code. It's like, who am I to say that "I don't like this code, let's refactor it to my liking". Of course, sometimes it warranted, and if the variables was to be like `i, x, c` - then yes, but here it's so minor I don't really think it's necessary. I do appreciate the thought to have more consistent code though, and maybe it's something we (I?) should weigh more, but at the end of the day we are multiple people working on the codebase, and a lot of time even the code I write is very inconsistent in regards to naming and other code-style (eg. `function` vs `const`). That's not to say I disregard it at all - but it's a balance of what is worth changing. But it's definitely something to keep in mind, and try to keep in check. 


> Coming back to this and idx specifically. I don't see why index makes it harder to keep track of variables? As mentioned, I wouldn't un-abbreviate everything (e.g. de, as it has an established meaning in the context of dhis2)

I didn't mean to comment on this example specifically. It was more in general. In this case I do agree with you; I do prefer the `index`-version. It was more that it feels so petty to me to change variable names in other people's code. Again, I might be weird that way. I probably have more against it if a change was to only rename variables - but here I did actually do some other changes, so it wouldn't be that unnatural to also rename them -  so I do understand your comment.